runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/client/grpc/GrpcCommitWorkStream.java

```diff
@@ -317,29 +350,71 @@ public boolean commitWorkItem(
       if (!canAccept(commitRequest.getSerializedSize() + computation.length())) {
         return false;
       }
-      PendingRequest request = new PendingRequest(computation, commitRequest, onDone);
+
+      PendingRequest request = PendingRequest.create(computation, commitRequest, onDone);
       add(idGenerator.incrementAndGet(), request);
       return true;
     }
 
     /** Flushes any pending work items to the wire. */
     @Override
     public void flush() {
-      flushInternal(queue);
-      queuedBytes = 0;
-      queue.clear();
+      try {
+        if (!isShutdown()) {
+          flushInternal(queue);
+        }
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      } finally {
+        queuedBytes = 0;
+        queue.clear();
+      }
     }
 
     void add(long id, PendingRequest request) {
-      assert (canAccept(request.getBytes()));
+      Preconditions.checkState(canAccept(request.getBytes()));
```

Comment:
it will throw if not true which will be handled by the AbstractWindmillStream#startStream method and in the StreamingEngineWorkCommitter (where we fail the commit and exit the loop)

Reply:
we call `request.onDone` with the error code when we shutdown and drain the queue
the check of isShutdown() in canAccept is to help us reject any more commits going into the stream if we are in a shutdown state

Reply:
What happens when we throw here? IIUC, the exception bubbles up to the commit loop in `StreamingEngineWorkCommitter` and gets suppressed there. It doesn't look like it'll call the commit callback with failure status, without it the work item will be in stuck state.

Reply:
i see what you mean now let me fix

Reply:
done